server/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -87,45 +86,61 @@ public static ByteSizeValue ofPb(long size) {
         return of(size, PB);
     }
 
-    private final long size;
-    private final ByteSizeUnit unit;
+    static ByteSizeValue newByteSizeValue(long sizeInBytes, ByteSizeUnit desiredUnit) {
+        // Peel off some common cases to avoid allocations
+        if (desiredUnit == BYTES) {
+            if (sizeInBytes == 0) {
+                return ZERO;
+            }
+            if (sizeInBytes == 1) {
+                return ONE;
+            }
+            if (sizeInBytes == -1) {
+                return MINUS_ONE;
+            }
+        }
+        if (sizeInBytes < 0) {
+            throw new IllegalArgumentException("Values less than -1 bytes are not supported: " + sizeInBytes);
+        }
+        return new ByteSizeValue(sizeInBytes, desiredUnit);
+    }
+
+    private final long sizeInBytes;
+    private final ByteSizeUnit desiredUnit;
 
     public static ByteSizeValue readFrom(StreamInput in) throws IOException {
         long size = in.readZLong();
         ByteSizeUnit unit = ByteSizeUnit.readFrom(in);
-        if (unit == BYTES) {
-            return ofBytes(size);
+        if (in.getTransportVersion().onOrAfter(BYTE_SIZE_VALUE_ALWAYS_USES_BYTES)) {
+            return newByteSizeValue(size, unit);
+        } else {
+            return of(size, unit);
         }
-        return of(size, unit);
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
-        out.writeZLong(size);
-        unit.writeTo(out);
+        if (out.getTransportVersion().onOrAfter(BYTE_SIZE_VALUE_ALWAYS_USES_BYTES)) {
+            out.writeZLong(sizeInBytes);
+        } else {
+            out.writeZLong(Math.divideExact(sizeInBytes, desiredUnit.toBytes(1)));
```

Comment:
Just checking, we need this because before BYTE_SIZE_VALUE_ALWAYS_USES_BYTES the "unit" was both the unit for "size" and the display unit, correct?

Reply:
Yes that's right. The motivation for this was so `ByteSizeValue` could survive a round trip via transport, and in the old protocol, `size` was an integer denominated in the given unit. Now that we want to support 2 decimal places, this protocol was unable to preserve fractional units, so @rjernst suggested changing the protocol so it always sends bytes.